File: DSA/Java/src/dsa/lib/math/nums/KthLargest.java
package dsa.lib.math.nums;

import java.util.PriorityQueue;

//LeetCode703
class KthLargest {
    private final PriorityQueue<Integer> queue;
    private final int k;

    public KthLargest(int k, int[] nums) {
        queue = new PriorityQueue<>();
        this.k = k;
        for (int num : nums) add(num);
    }

    public int add(int val) {
        queue.add(val);
        if (queue.size() > this.k) queue.poll();
        return queue.peek();
    }
}

File: DSA/Python/src/dsa/lib/math/nums/kth_largest.py
import heapq
from typing import List

from dsa.container.lists.queues import PriorityQueue


# LeetCode703
class KthLargest1:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.queue = heapq.nlargest(k, nums)
        heapq.heapify(self.queue)

    def add(self, val: int) -> int:
        heapq.heappush(self.queue, val)
        if len(self.queue) > self.k: heapq.heappop(self.queue)
        return self.queue[0]


class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.queue: PriorityQueue = PriorityQueue()
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        self.queue.push(val)
        if self.queue.size() > self.k:
            self.queue.pop()
        return self.queue.peek()
